Lib/zipimport.py

```diff
@@ -88,12 +88,9 @@ def __init__(self, path):
                     raise ZipImportError('not a Zip file', path=path)
                 break
 
-        try:
-            files = _zip_directory_cache[path]
-        except KeyError:
-            files = _read_directory(path)
-            _zip_directory_cache[path] = files
-        self._files = files
+        if path not in _zip_directory_cache:
+            _zip_directory_cache[path] = _read_directory(path)
+        self._cache_is_valid = True
```

Comment:
Why the separate key instead of setting `self._files` to `None` or deleting the entry from `_zip_directory_cache` to signal the cache needs to be updated?

Reply:
Thanks for catching that @brettcannon. I wrote a small test case for the scenario that you described and using the poisoned cache did turn out to be an issue. (On a side note, this problem also existed in the original design because `_files` was being used to cache the entry in `_zip_directory_cache`, and this "cache of a cache" creates a similar cache coherence issue).

I removed the extra flag and opted to delete the entry from `_zip_directory_cache` on cache invalidation. Please let me know if this works for you